.circleci/config.yml

```diff
@@ -1,117 +1,92 @@
-restore_cache: &restore_cache
-  restore_cache:
-    # If there isn't a match to the first key, it'll do a partial match of the
-    # second. That means after the first cache save there will always be a cache
-    # hit, but it might be an older version of the cache
-    keys:
-      - yarn-deps-{{ checksum "yarn.lock" }}
-      - yarn-deps-
-
-# By default when yarn runs it does not check the filesystem to ensure the
-# packages it expects to be installed are actually installed. Using the
-# --check-files flag ensures that any packages or files missing or out of date
-# on the file system (i.e. those which might be restored from cache) match what
-# the yarn.lock file specifies
-verify_dependencies: &verify_dependencies
-  run: yarn --check-files
-
-filter_master: &filter_master
-  filters:
-    branches:
-      only:
-        - master
-
-version: 2
 jobs:
   relay:
-    <<: *setup_workspace
+    executor: node/build
     steps:
-      - checkout
-      - <<: *restore_cache
-      - <<: *verify_dependencies
+      - yarn/setup
       - run: yarn relay
   lint:
-    <<: *setup_workspace
+    executor: node/build
     steps:
-      - checkout
-      - <<: *restore_cache
-      - <<: *verify_dependencies
+      - yarn/setup
       - run: yarn lint
   type-check:
-    <<: *setup_workspace
+    executor: node/build
     steps:
-      - checkout
-      - <<: *restore_cache
-      - <<: *verify_dependencies
+      - yarn/setup
       - run: yarn type-check
-
   deploy:
-    <<: *setup_workspace
+    executor: node/build
+    # Deploy doesn't use yarn/setup intentially. It's assumed that update-cache
+    # has already ran by the time deploy executes so the dependencies in the
+    # cache should be up-to-date therefore not requiring check_and_install to be
+    # executed
     steps:
-      # Note that the deploy job doesn't call verify_dependencies. It's this
-      # way for two reasons:
-      #   1) It waits for the update-cache job to run so it's guarenteed that
-      #      the cache will contain the packages needed for this deployment
-      #      process. (i.e. there's not reason to run yarn install)
-      #   2) When npm/yarn install runs it executes the pre and post install
-      #      hooks of dependencies which means child packages could read any
-      #      environment variables available. While this doesn't 100% protect
-      #      us, it does reduce that attack surface. Note that this job is the
-      #      _only_ job with the environment variables needed to publish thanks
-      #      to the context being specified in the workflow.
       - checkout
-      - <<: *restore_cache
+      - yarn/load_dependencies
       # Setup the .npmrc with the proper registry and auth token to publish
       - run: echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" >> ~/.npmrc
       - run: yarn semantic-release
+  pre-deploy:
+    executor: node/build
+    steps:
+      - queue/until_front_of_line:
```

Comment:
Interesting. We could use something similar to prevent builds and pushes of new `staging` docker images while a `production` release (which does a promote of the `staging` image) is pending.